lib/elixir/lib/map.ex

```diff
@@ -93,11 +96,10 @@ defmodule Map do
   """
   @spec new(Enum.t, (term -> {key, value})) :: map
   def new(enumerable, transform) do
-    fun = fn el, acc ->
-      {k, v} = transform.(el)
-      put(acc, k, v)
-    end
-    Enum.reduce(enumerable, %{}, fun)
+    enumerable
+    |> Enum.reduce([], &[transform.(&1) | &2])
+    |> :lists.reverse
+    |> :maps.from_list
```

Comment:
In particular, `Enum.into/2` should optimize into a single :maps.from_list call if the first argument is a list and the second is a map (not a struct). I believe we already do it today but, if we don't, we certainly should.